sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionBase.scala

```diff
@@ -326,6 +326,18 @@ abstract class TypeCoercionBase extends TypeCoercionHelper {
     }
   }
 
+  object CastCoercion extends TypeCoercionRule {
+
+    override val transform: PartialFunction[Expression, Expression] = {
+      case c @ Cast(child, CharType(_) | VarcharType(_), _, _) if child.resolved =>
+        child.dataType match {
+          case _: StringType => c
+          case _ => c.withNewChildren(Seq(Cast(child, StringType)))
```

Comment:
Why is this necessary? We should support directly cast int to varchar for example. The `Cast` should have a general `castToString` method that supports all string types.